### Data Merging
#### The existing file is in bits and needs to be merged at once using Python.

In [1]:
import pandas as pd

In [3]:
import glob

# Set the directory where the Excel files are located
path = 'C:/Users/Akon.usoh/Desktop/PERSONAL/PROJECTS MI/LOAN PORTFOLIO MANAGEMENT'

# Use glob to find all Excel files in the directory
all_files = glob.glob(path + "/*.xlsx")

# Create an empty list to hold the dataframes
dfs = []

# Loop through each Excel file and read it into a dataframe
for filename in all_files:
    df = pd.read_excel(filename, index_col=None, header=0)
    dfs.append(df)

# Concatenate all the dataframes into one dataframe
merged_df = pd.concat(dfs, axis=0, ignore_index=True)

# Save the merged dataframe to a new Excel file
merged_df.to_excel('C:/Users/Akon.usoh/Desktop/PERSONAL/PROJECTS MI/LOAN PORTFOLIO MANAGEMENT/loan_data.xlsx', index=False)

### Load the data: 
#### Load the data into a pandas dataframe and inspect the data for any missing values, duplicates, or inconsistent data.

In [5]:
merged_df.head()

name  gender                           email_address  \
0    Andie Thorpe  Female        Andie_Thorpe584334985@kyb7t.site   
1  Savannah Moore  Female      Savannah_Moore1139609715@ds59r.net   
2   Christy Locke  Female  Christy_Locke1627978587@nb44i.software   
3       Maia Owen  Female     Maia_Owen1372614818@cdpfn.solutions   
4     Rick Briggs    Male       Rick_Briggs1385538007@nanoff.club   

     phone_number  credit_score  monthly_debt  monthly_income  loan_amount  \
0  5-454-870-2512         628.0       11691.0          4398.0      65705.0   
1  7-256-352-8173         403.0        7979.0          1883.0      28865.0   
2  6-463-821-2804         753.0       19651.0          5883.0      40558.0   
3  7-147-877-5334         539.0       18641.0          5788.0      82146.0   
4  2-483-677-2032         582.0       19367.0          4426.0      71698.0   

   property_value  employment_length  credit_card_balance  credit_limit  \
0         41806.0                6.0              45016.0       71212.0   
1         80383.0                5.0              63646.0       43392.0   
2         64421.0                4.0              13172.0       28881.0   
3         39600.0                5.0               4851.0       81440.0   
4        102721.0                0.0              94346.0       69485.0   

   loan_status  
0          0.0  
1          0.0  
2          0.0  
3          0.0  
4          0.0

In [4]:
merged_df.shape

(900009, 13)

In [6]:
df = merged_df.copy()
#Make a copy of the data for replication purpose

In [7]:
# Inspect the data
print(df.tail())
print(df.info())

                                                     name  gender  \
900004                                   Charlotte Slater  Female   
900005                                     Lindsay Graham  Female   
900006                                  Rosemary Marshall  Female   
900007                                    Colleen Widdows  Female   
900008  data generated with OnlineDataGenerator availa...     NaN   

                                email_address    phone_number  credit_score  \
900004    Charlotte_Slater495202665@lhp4j.pro  6-711-443-2350         794.0   
900005    Lindsay_Graham923405452@ohqqh.video  4-532-028-1652         822.0   
900006  Rosemary_Marshall1534257889@3wbkp.edu  2-710-631-8682         324.0   
900007     Colleen_Widdows35833202@ohqqh.name  4-131-566-6344         597.0   
900008                                    NaN             NaN           NaN   

        monthly_debt  monthly_income  loan_amount  property_value  \
900004       19638.0          9995.0     

### Clean the data: 
#### Clean the data by handling missing values, duplicates, and inconsistent data.

In [8]:
# Handle missing values
df = df.dropna()

In [9]:
# Handle duplicates
df = df.drop_duplicates()

In [10]:
# Handle inconsistent data
df['employment_length'] = df['employment_length'].replace('10+', '10')
df['employment_length'] = df['employment_length'].replace('< 1', '0')

In [11]:
# Convert the float columns to integer
df['credit_score'] = df['credit_score'].astype(int)
df['monthly_debt'] = df['monthly_debt'].astype(int)
df['monthly_income'] = df['monthly_income'].astype(int)
df['loan_amount'] = df['loan_amount'].astype(int)
df['property_value'] = df['property_value'].astype(int)
df['employment_length'] = df['employment_length'].astype(int)
df['credit_card_balance'] = df['credit_card_balance'].astype(int)
df['credit_limit'] = df['credit_limit'].astype(int)
df['loan_status'] = df['loan_status'].astype(int)

In [12]:
df.sample(5)
#Check the cleaned data

name  gender  \
852169       Karla Daniells  Female   
833652           Elena Rees  Female   
523173  Cristal Collingwood  Female   
751364           Rhea Hobbs  Female   
594796        Caitlyn Allen  Female   

                                     email_address    phone_number  \
852169       Karla_Daniells933858121@bqkv0.website  8-138-747-7576   
833652            Elena_Rees1517305161@xqj6f.autos  1-456-513-1504   
523173  Cristal_Collingwood1128965221@dbxli.center  0-126-401-2654   
751364               Rhea_Hobbs255157126@evyvh.pro  8-105-546-6148   
594796        Caitlyn_Allen1788550190@bu2lo.center  1-825-163-0674   

        credit_score  monthly_debt  monthly_income  loan_amount  \
852169           735          3441           13054        54404   
833652           311          9779            8972        82214   
523173           513          6570           14581        55275   
751364           644          5362            7052        44919   
594796           682          3878            4728        84005   

        property_value  employment_length  credit_card_balance  credit_limit  \
852169           53371                  9                66271         36123   
833652           90318                  0                35293         25066   
523173           51582                  1                77369         98667   
751364           82973                 10                75563         73831   
594796          111999                  7                87829         90322   

        loan_status  
852169            0  
833652            0  
523173            1  
751364            1  
594796            1

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 896000 entries, 0 to 900007
Data columns (total 13 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   name                 896000 non-null  object
 1   gender               896000 non-null  object
 2   email_address        896000 non-null  object
 3   phone_number         896000 non-null  object
 4   credit_score         896000 non-null  int32 
 5   monthly_debt         896000 non-null  int32 
 6   monthly_income       896000 non-null  int32 
 7   loan_amount          896000 non-null  int32 
 8   property_value       896000 non-null  int32 
 9   employment_length    896000 non-null  int32 
 10  credit_card_balance  896000 non-null  int32 
 11  credit_limit         896000 non-null  int32 
 12  loan_status          896000 non-null  int32 
dtypes: int32(9), object(4)
memory usage: 64.9+ MB


### Feature engineering: 
#### Create new features from existing features that may help improve the model's performance.

In [14]:
# Create a new feature to calculate the debt-to-income ratio
df['debt_to_income_ratio'] = df['monthly_debt'] / df['monthly_income']

#### The debt-to-income (DTI)

DTI ratio is a financial measure that compares an individual's monthly debt payments to their monthly gross income. It is used to assess an individual's ability to manage monthly debt payments and make timely payments on new debts.

In the context of loan portfolio management, the DTI ratio is an important metric used by lenders to evaluate a borrower's creditworthiness and ability to repay a loan. A high DTI ratio indicates that a borrower has a significant amount of debt relative to their income, which may make it difficult for them to make timely payments on a new loan.

In the example provided earlier, we created a new feature to calculate the debt-to-income ratio using the borrower's monthly debt and monthly income. This feature may help improve the predictive power of the model by providing additional information about the borrower's financial health and ability to repay the loan. A low DTI ratio indicates that a borrower has a low level of debt relative to their income, which may make them a more attractive candidate for a loan. On the other hand, a high DTI ratio may indicate that a borrower has a high level of debt relative to their income and may be at a higher risk of defaulting on the loan.

In [15]:
# Create a new feature to calculate the loan-to-value ratio
df['loan_to_value_ratio'] = df['loan_amount'] / df['property_value']

#### The loan-to-value (LTV) 

LTV ratio is a financial measure that compares the amount of a loan to the value of the asset being purchased with the loan. In the context of loan portfolio management, the LTV ratio is used by lenders to evaluate the risk associated with a loan and to determine the maximum amount of financing they are willing to provide.

A high LTV ratio indicates that the borrower is seeking a loan that represents a significant portion of the value of the asset being purchased. For example, if a borrower is seeking a $100,000 loan to purchase a $150,000 property, the LTV ratio would be 0.67, or 67%. A high LTV ratio may indicate that the borrower has a limited amount of equity in the asset being purchased, which may make it more difficult for them to repay the loan in the event of default.

In the example provided earlier, we created a new feature to calculate the loan-to-value ratio using the loan amount and property value. This feature may help improve the predictive power of the model by providing additional information about the level of risk associated with the loan. A high LTV ratio may indicate that the borrower is at a higher risk of defaulting on the loan, while a low LTV ratio may indicate that the borrower has a larger equity stake in the asset being purchased and may be at a lower risk of defaulting on the loan.

In [16]:
# Create a new feature to calculate the credit utilization ratio
df['credit_utilization_ratio'] = df['credit_card_balance'] / df['credit_limit']

#### The credit utilization ratio

The credit utilization ratio is a financial measure that compares a borrower's credit card balances to their credit card limits. In other words, it is the percentage of the total available credit that a borrower is currently using.

In the context of loan portfolio management, the credit utilization ratio is used by lenders to evaluate a borrower's creditworthiness and to assess the risk associated with extending credit. A high credit utilization ratio may indicate that a borrower is relying heavily on credit and may be at a higher risk of defaulting on a loan.

In the example provided earlier, we created a new feature to calculate the credit utilization ratio using the borrower's credit card balance and credit limit. This feature may help improve the predictive power of the model by providing additional information about the borrower's creditworthiness and ability to manage their finances. A low credit utilization ratio indicates that a borrower is using a relatively small percentage of their available credit, which may make them a more attractive candidate for a loan. On the other hand, a high credit utilization ratio may indicate that a borrower is relying heavily on credit and may be at a higher risk of defaulting on the loan.

### Data modeling: 
#### Build a machine learning model to predict loan default.

In [17]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [18]:
# Check the skewness of the data
print("Skewness of data:", df.skew())

Skewness of data: credit_score               -0.000780
monthly_debt               -0.001329
monthly_income             -0.001090
loan_amount                 0.000271
property_value             -0.002662
employment_length          -0.001323
credit_card_balance        -0.001522
credit_limit                0.000259
loan_status                -0.233723
debt_to_income_ratio        3.341964
loan_to_value_ratio         1.694953
credit_utilization_ratio    3.367913
dtype: float64


C:\Users\Akon.usoh\AppData\Local\Temp\ipykernel_20820\2017531126.py:2: FutureWarning: Dropping of nuisance columns in DataFrame reductions (with 'numeric_only=None') is deprecated; in a future version this will raise TypeError.  Select only valid columns before calling the reduction.
  print("Skewness of data:", df.skew())


Based on the provided information, the columns "debt_to_income_ratio" and "credit_utilization_ratio" have a skewness of 3.34 and 3.37, respectively, which indicates that the data is highly skewed. In such cases, using median is generally better than using mean, as mean can be heavily influenced by the extreme values in the skewed distribution. Therefore, in this case, using median would be a better option.

In [19]:
# Split the data into training and testing sets
X = df[['credit_score', 'debt_to_income_ratio', 'loan_to_value_ratio', 'credit_utilization_ratio']]
y = df['loan_status']

In [20]:
# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

In [21]:
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

In [22]:
# Define the preprocessor
preprocessor = ColumnTransformer([
    ('num_imputer', SimpleImputer(strategy='median'), ['credit_score', 'debt_to_income_ratio', 'loan_to_value_ratio', 'credit_utilization_ratio'])
])

In [23]:
# Fit the preprocessor on the training set
preprocessor.fit(X_train)

ColumnTransformer(transformers=[('num_imputer',
                                 SimpleImputer(strategy='median'),
                                 ['credit_score', 'debt_to_income_ratio',
                                  'loan_to_value_ratio',
                                  'credit_utilization_ratio'])])

In [24]:
# Transform the training and test sets
X_train_transformed = preprocessor.transform(X_train)
X_test_transformed = preprocessor.transform(X_test)

#### Random Forest Classifier
Here, we use the preprocessor to transform the training set using the preprocessor.transform method. 

Train a Random Forest Classifier model on the transformed data set and make predictions using the predict method. 

Aside this, other Model we can attempt for Loan Management Portfolio includes DecisionTreeClassifier, SVM, Logistic regression.

In [25]:
from sklearn.ensemble import RandomForestClassifier
# Train a random forest classifier on the transformed training set
rf = RandomForestClassifier(n_estimators=100, random_state=42)
rf.fit(X_train_transformed, y_train)

RandomForestClassifier(random_state=42)

In [26]:
# Predict the churn labels for the whole dataset
X_transformed = preprocessor.transform(X)
y_pred = rf.predict(X_transformed)

In [27]:
df['loan_prediction'] = y_pred

In [28]:
df.head()

name  gender                           email_address  \
0    Andie Thorpe  Female        Andie_Thorpe584334985@kyb7t.site   
1  Savannah Moore  Female      Savannah_Moore1139609715@ds59r.net   
2   Christy Locke  Female  Christy_Locke1627978587@nb44i.software   
3       Maia Owen  Female     Maia_Owen1372614818@cdpfn.solutions   
4     Rick Briggs    Male       Rick_Briggs1385538007@nanoff.club   

     phone_number  credit_score  monthly_debt  monthly_income  loan_amount  \
0  5-454-870-2512           628         11691            4398        65705   
1  7-256-352-8173           403          7979            1883        28865   
2  6-463-821-2804           753         19651            5883        40558   
3  7-147-877-5334           539         18641            5788        82146   
4  2-483-677-2032           582         19367            4426        71698   

   property_value  employment_length  credit_card_balance  credit_limit  \
0           41806                  6                45016         71212   
1           80383                  5                63646         43392   
2           64421                  4                13172         28881   
3           39600                  5                 4851         81440   
4          102721                  0                94346         69485   

   loan_status  debt_to_income_ratio  loan_to_value_ratio  \
0            0              2.658254             1.571664   
1            0              4.237387             0.359093   
2            0              3.340303             0.629577   
3            0              3.220629             2.074394   
4            0              4.375734             0.697988   

   credit_utilization_ratio  loan_prediction  
0                  0.632141                1  
1                  1.466768                0  
2                  0.456078                0  
3                  0.059565                0  
4                  1.357789                0

In [29]:
df.isnull().sum()

name                        0
gender                      0
email_address               0
phone_number                0
credit_score                0
monthly_debt                0
monthly_income              0
loan_amount                 0
property_value              0
employment_length           0
credit_card_balance         0
credit_limit                0
loan_status                 0
debt_to_income_ratio        0
loan_to_value_ratio         0
credit_utilization_ratio    0
loan_prediction             0
dtype: int64

In the above, if the column 'Loan_Prediction' contains only 0 or 1 values, it represents a binary classification output where 0 and 1 are used to represent the predicted class labels for each observation in the dataset.

Typically, a value of 0 in this column represent one class, which is the negative class, and a value of 1 represent the other class, which is the positive class.

In [30]:
# Define the mapping from old cluster labels to new ones
cluster_mapping = {0: 'default_probability', 1: 'repayment_probability'}

# Apply the mapping to the cluster labels using the map() function
df['predicted_loan_status'] = df['loan_prediction'].map(cluster_mapping)

In [31]:
df.sample(10)

name  gender                           email_address  \
203616     Sharon Ulyatt  Female       Sharon_Ulyatt1869596263@zynuu.edu   
845468  Marissa Aldridge  Female  Marissa_Aldridge1019555186@avn7d.media   
874092    Phoebe Tindall  Female     Phoebe_Tindall1061820452@qu9ml.info   
414209   Irene Underhill  Female    Irene_Underhill1139533961@iaart.page   
47342       Rhea Allwood  Female    Rhea_Allwood1786186237@y96lx.website   
576182    Shannon Harris  Female   Shannon_Harris808466034@evyvh.digital   
9155        Marvin Lloyd    Male       Marvin_Lloyd1237366525@voylg.page   
58520         Sage Yoman  Female          Sage_Yoman1105672786@lyvnc.biz   
456529      Aeris Roscoe  Female      Aeris_Roscoe1008230220@dbxli.store   
430255      Raquel Jones  Female    Raquel_Jones1791813418@ohqqh.auction   

          phone_number  credit_score  monthly_debt  monthly_income  \
203616  4-645-156-2774           775          4852            5410   
845468  5-152-051-1618           551         10735            8753   
874092  5-543-857-6531           588          5241            5819   
414209  1-325-826-3768           593          6802           18907   
47342   2-268-655-7146           474         19676            2573   
576182  6-877-541-7670           707         16386           11450   
9155    4-782-847-1455           383          6970           11063   
58520   7-388-375-4484           846         18363            4098   
456529  8-655-054-1607           520          2562            7960   
430255  0-704-681-7541           835          2386           10202   

        loan_amount  property_value  employment_length  credit_card_balance  \
203616        46624          113561                  3                10407   
845468        34941           95402                  4                38844   
874092        85741           60074                  5                 3642   
414209        79799          115304                  3                11351   
47342         56945           42602                 10                84967   
576182        54792           53168                  1                33908   
9155          65821           20622                  8                20301   
58520         98309           81569                  2                15561   
456529        62779          117846                  2                38459   
430255        35939           30680                  4                41228   

        credit_limit  loan_status  debt_to_income_ratio  loan_to_value_ratio  \
203616         60101            0              0.896858             0.410563   
845468         23127            0              1.226437             0.366250   
874092         35119            0              0.900670             1.427256   
414209         74478            1              0.359761             0.692075   
47342          31286            0              7.647105             1.336674   
576182         84149            1              1.431092             1.030545   
9155           62963            0              0.630028             3.191785   
58520          50980            0              4.480966             1.205225   
456529         81613            1              0.321859             0.532721   
430255         36633            1              0.233876             1.171415   

        credit_utilization_ratio  loan_prediction  predicted_loan_status  
203616                  0.173159                0    default_probability  
845468                  1.679595                0    default_probability  
874092                  0.103705                0    default_probability  
414209                  0.152407                1  repayment_probability  
47342                   2.715815                0    default_probability  
576182                  0.402952                1  repayment_probability  
9155                    0.322427                0    default_probability  
58520                   0.305237                0    default_

In [32]:
df1 = df.copy()

In [33]:
# Save the cleaned data to a new CSV file
df.to_csv('loan_portfolio_mgt.csv', index=False)

In [36]:
# Closing remark
print("Thank you for going through my loan management portfolio!")
print("If you have any questions or suggestions, please feel free to reach out to me on LinkedIn at https://linkedin.com/in/akonusoh")

Thank you for going through my loan management portfolio!
If you have any questions or suggestions, please feel free to reach out to me on LinkedIn at https://linkedin.com/in/akonusoh
